pkg/tsdb/stackdriver/stackdriver.go

```diff
@@ -351,40 +351,64 @@ func (e *StackdriverExecutor) unmarshalResponse(res *http.Response) (Stackdriver
 }
 
 func (e *StackdriverExecutor) parseResponse(queryRes *tsdb.QueryResult, data StackdriverResponse, query *StackdriverQuery) error {
-	metricLabels := make(map[string][]string)
-	resourceLabels := make(map[string][]string)
-	var resourceTypes []string
-
-	for _, series := range data.TimeSeries {
-		if !containsLabel(resourceTypes, series.Resource.Type) {
-			resourceTypes = append(resourceTypes, series.Resource.Type)
-		}
-	}
+	labels := make(map[string][]string)
+	labels["resource.type"] = make([]string, 0)
 
 	for _, series := range data.TimeSeries {
 		points := make([]tsdb.TimePoint, 0)
+		seriesLabels := make(map[string]string)
 
 		defaultMetricName := series.Metric.Type
-		if len(resourceTypes) > 1 {
-			defaultMetricName += " " + series.Resource.Type
+
+		seriesLabels["resource.type"] = series.Resource.Type
+		if !containsLabel(labels["resource.type"], series.Resource.Type) {
+			labels["resource.type"] = append(labels["resource.type"], series.Resource.Type)
 		}
 
 		for key, value := range series.Metric.Labels {
-			if !containsLabel(metricLabels[key], value) {
-				metricLabels[key] = append(metricLabels[key], value)
+			if !containsLabel(labels["metric.label."+key], value) {
+				labels["metric.label."+key] = append(labels["metric.label."+key], value)
 			}
 			if len(query.GroupBys) == 0 || containsLabel(query.GroupBys, "metric.label."+key) {
 				defaultMetricName += " " + value
 			}
+			seriesLabels["metric.label."+key] = value
 		}
 
 		for key, value := range series.Resource.Labels {
-			if !containsLabel(resourceLabels[key], value) {
-				resourceLabels[key] = append(resourceLabels[key], value)
-			}
 			if containsLabel(query.GroupBys, "resource.label."+key) {
 				defaultMetricName += " " + value
 			}
+
+			if !containsLabel(labels["resource.label."+key], value) {
+				labels["resource.label."+key] = append(labels["resource.label."+key], value)
+			}
+			seriesLabels["resource.label."+key] = value
+		}
+
+		for labelType, labelTypeValues := range series.MetaData {
+			for labelKey, labelValue := range labelTypeValues {
+				output := []string{}
+				if values, ok := labelValue.([]interface{}); ok {
+					for _, v := range values {
+						output = append(output, v.(string))
+					}
+				} else if boolValue, ok := labelValue.(bool); ok {
+					output = append(output, strconv.FormatBool(boolValue))
+				} else if strValue, ok := labelValue.(string); ok {
+					output = append(output, strValue)
+				}
+
+				for _, value := range output {
+					key := toSnakeCase(fmt.Sprintf("metadata.%s.%s", labelType, labelKey))
+
+					if !containsLabel(labels[key], value) {
+						labels[key] = append(labels[key], value)
+					}
+
+					seriesLabels[key] = value
```

Comment:
No that will just append the value to the output array. [This](https://github.com/grafana/grafana/blob/acb61a6a3561d346f7e322834864306e8d5ad5a4/pkg/tsdb/stackdriver/stackdriver_test.go#L434-L442) unit test should validate that behavior.

Reply:
For `labels[key]`, yes, which is why I didn't comment on that loop. However, for `seriesLabels[key]`, which becomes part of the legend, it looks like only the last value will be recorded....

Reply:
Ah I see, good catch. Did some more refactoring. In case metadata.system_labels.x is a string array the values will be joined with a comma. 

`metadata.system_labels.name = ["value1", "value2"]`
when used in the alias by field like this
`{{metadata.system_labels.name}}` 
the legend will be named `value1, value2`